### ISSUE TITLE
Pychess svn build crashes on start at the engine search

### ISSUE THREAD

**Issue by pychess** [pychess/pychess#424]
Original [issue 424](https://code.google.com/p/pychess/issues/detail?id=424) reported by [maxmusterm](https://code.google.com/u/114206061950813382165/) 2009-03-22

**What steps will reproduce the problem?**
**1.**
install gnuchess and another engine (I say gnuchess because it only happens
with gnuchess on my pc)
**2.**
get the newest subversion build
**3.**
try to start pychess

**What is the expected output? What do you see instead?**
The expected output is nothing but I see

22:46:52 (u'Gnuchess', '22:03:52.379') Error: Success
22:46:53 (u'Crafty', '22:03:52.933') Error: Permission denied

/home/blacklotus/work/pychess/lib/pychess/System/SubProcess.py:82: Warning:
Invalid file descriptor.

  channel.close()
  File "./pychess", line 40, in <module>
    import pychess.Main
  File "/home/blacklotus/work/pychess/lib/pychess/Main.py", line 17, in
<module>
    from pychess.widgets import tipOfTheDay
  File "/home/blacklotus/work/pychess/lib/pychess/widgets/tipOfTheDay.py",
line 7, in <module>
    uistuff.cacheGladefile("tipoftheday.glade")
  File "/home/blacklotus/work/pychess/lib/pychess/System/uistuff.py", line
353, in cacheGladefile
    pool.start(readit)
  File "/home/blacklotus/work/pychess/lib/pychess/Players/engineNest.py",
line 344, in run
    rechecks = self._findOutMore(toBeDiscovered)
  File "/home/blacklotus/work/pychess/lib/pychess/Players/engineNest.py",
line 269, in _findOutMore
    engine.start()
  File "/home/blacklotus/work/pychess/lib/pychess/Players/CECPEngine.py",
line 145, in start
    self.__startBlocking()
  File "/home/blacklotus/work/pychess/lib/pychess/Players/CECPEngine.py",
line 158, in __startBlocking
    assert r == "ready"
<type 'exceptions.AssertionError'> 

**Does it happen every time?**
Yes it does.

**Please provide any additional information below.**
A workaround is to uninstall gnuchess, start pychess, close pychess,install
gnuchess and then start pychess again.
The cause of the error is that pychess is "to fast" and it doesn't find the
childprocess of gnuchess in time to close the channels. (i know i'm
horrible at explaining)
The easiest code fix (I found in the few minutes i looked over the code
there sure is a better fix) would be a sleep in SubProcess.py in line 80 like:

```
    time.sleep(0.1)
    for channel in (self.inChannel, self.outChannel, self.errChannel):
```

That fixed it for me.

**Comment by pychess** [comment]
Original [comment 1.](https://code.google.com/p/pychess/issues/detail?id=424#c1) posted by [lobais](https://code.google.com/u/lobais/) 2009-03-22

It would seam to me, that the problem had to do with Crafty (Error: Permission
denied)? Did you try to set the +x flag on /usr/bin/crafty, or where it lies?

**Comment by pychess** [comment]
Original [comment 3.](https://code.google.com/p/pychess/issues/detail?id=424#c3) posted by [maxmusterm](https://code.google.com/u/114206061950813382165/) 2009-03-22

You misunderstood it.
With the sleep of 0.1 it worked.
__closeChannels fails for me because channel is set wrong don't ask me how this
exactly can happen maybe my pc is simply to slow but like i said the fix works like
this for me and waiting 0.1 sec is no big deal is it?
What I did:
I removed crafty - result: the next engine in list did not work
I removed gnuchess - result: all worked fine
I changed the code with the sleep it worked without it didn't.

**Comment by pychess** [comment]
Original [comment 4.](https://code.google.com/p/pychess/issues/detail?id=424#c4) posted by [lobais](https://code.google.com/u/lobais/) 2009-03-22

I'm sorry about the misunderstanding.

For the solution you are right, that 0.1s is not such a big deal, but it would be
much prefferable, if we can find the 'actual' problem, so even slower computers than
yours will work as well, and people with 20 engines wont have to wait more than 2s on
startup.
The problem may very well have to do with the channels, which have given trouble
before. The question is where the race lies, what has to be done before closing them,
that takes time?

If you want to, feel very free to take a deeper breath into the code.
Anyway, thanks a lot for your help already.

**Comment by pychess** [comment]
Original [comment 5.](https://code.google.com/p/pychess/issues/detail?id=424#c5) posted by [gbtami](https://code.google.com/u/gbtami/) 2009-03-23
- added label: Defect

**Comment by pychess** [comment]
Original [comment 6.](https://code.google.com/p/pychess/issues/detail?id=424#c6) posted by [mattgatto](https://code.google.com/u/mattgatto/) 2009-06-21

I'm pretty sure this is a duplicate of bug 337.
- changed status to: Duplicate
